src/gausskernel/storage/access/k2/pg_gate_api.cpp

```diff
@@ -24,59 +24,109 @@ Copyright(c) 2022 Futurewei Cloud
 // When we mix certain C++ standard lib code and pg code there seems to be a macro conflict that
 // will cause compiler errors in libintl.h. Including as the first thing fixes this.
 #include <libintl.h>
+#include <unordered_map>
+#include <assert.h>
+#include <atomic>
+#include <memory>
 
 #include "access/k2/pg_gate_api.h"
+#include "access/k2/pg_memctx.h"
 
 #include "k2pg-internal.h"
 #include "session.h"
 #include "access/k2/k2pg_util.h"
 
 #include "utils/elog.h"
+#include "utils/errcodes.h"
 #include "pg_gate_defaults.h"
 #include "pg_gate_thread_local.h"
 #include "catalog/sql_catalog_client.h"
 
-#include <atomic>
-
 using namespace k2pg::gate;
 
 namespace {
-// Using a raw pointer here to fully control object initialization and destruction.
-// k2pg::gate::PgGateApiImpl* api_impl;
-std::atomic<bool> api_impl_shutdown_done;
 
-} // anonymous namespace
+    class PgGate {
+        public:
+        // TODO: remove K2PgTypeEntity to map to use type oid to map to k2 type directly
+        PgGate(const K2PgTypeEntity *k2PgDataTypeArray, int count, K2PgCallbacks callbacks) {
+            // Setup type mapping.
+            for (int idx = 0; idx < count; idx++) {
+                const K2PgTypeEntity *type_entity = &k2PgDataTypeArray[idx];
+                type_map_[type_entity->type_oid] = type_entity;
+            }
+            catalog_manager_ = std::make_shared<k2pg::catalog::SqlCatalogManager>();
+            catalog_client_ = std::make_shared<k2pg::catalog::SqlCatalogClient>(catalog_manager_);
+            catalog_manager_->Start();
+            initialized_ = true;
+        };
+
+        ~PgGate() {
+              catalog_manager_->Shutdown();
+              initialized_ = false;
+        };
+
+        bool IsInitialized() {
```

Comment:
I don't think the initialized flag gets us anything since you can't construct this object without initializing it anyway. I think we can just assume that if constructed,  it is initialized

Reply:
make sense with the current implementation. Removed the initialized flag as suggested.